A36-xds-for-servers.md

```diff
@@ -193,8 +191,43 @@ required. This is to allow implementations that only support L4 xDS features to
 avoid L7 plumbing and implementation. This has no impact on the resource
 validation and NACKing behavior called for in A39.
 
+If an XdsServer implementation uses RouteConfiguration or supports any HTTP
+filters other than the hard-coded Router, then
+`HttpConnectionManager.route_config` and `HttpConnectionManager.rds` must be
+supported and RouteConfigurations must be validated. RouteConfiguration
+validation logic inherits all previous validations made for client-side usage
+as RDS does not distinguish between client-side and server-side. That is
+predomenently defined in [gRFC A28][A28-validation], although note that
+configuration for all VirtualHosts have been validated on client-side since
+sharing the XdsClient was introduced, yet was not documented in a gRFC. The
+validation must be updated to allow [Route.non_forwarding_action][] as a valid
+`action`. The VirtualHost is selected on a per-RPC basis using the RPC's
```

Comment:
> it might make more sense to simply tell the server what it's name is

You mean xds would tell us what authority to use and we'd overwrite what was in the request? That sounds like a substantial architecture change and would impact control plane configuration APIs as well.

> The only benefit I can see to determining this dynamically on the gRPC server is to support some form of virtual hosting, which doesn't seem like something that most gRPC users would care about.

FWIW, I doubt many REST-style servers sitting behind an Envoy sidecar care as well. Although, in REST the hostname still matters without virtual hosting because you need to build URLs in responses.

> and the overhead of determining this dynamically is probably not huge

The overhead may be very small, since we'd generally expect a single entry that was a wildcard. Since we have to go through the rest of route selection per-RPC _anyway_, it doesn't add any complexity either. We could pay some complexity to optimize the `*` case in the code and short-circuit the virtual host selection. I'd much rather optimize-it-out in the common case than make an architectural change.

I will note that in not-_that_-uncommon situations the hostname can vary per request. When you have a reverse proxy sitting in front of a backend it can end up putting multiple different clients' requests on the same connection. The moment there are two ways to refer to a service (e.g., global vs regional name) you see this change per-request.

Reply:
> You mean xds would tell us what authority to use and we'd overwrite what was in the request? That sounds like a substantial architecture change and would impact control plane configuration APIs as well.

I was thinking more along the lines of the application telling the server what its name is via the `XdsServerBuilder` API, so that the server can just ignore any non-matching `VirtualHost` that it gets from xDS.

I'm not really in favor of that option; I just wanted to make sure that we'd considering the alternatives.

> The overhead may be very small, since we'd generally expect a single entry that was a wildcard. Since we have to go through the rest of route selection per-RPC _anyway_, it doesn't add any complexity either. We could pay some complexity to optimize the `*` case in the code and short-circuit the virtual host selection. I'd much rather optimize-it-out in the common case than make an architectural change.
> 
> I will note that in not-_that_-uncommon situations the hostname can vary per request. When you have a reverse proxy sitting in front of a backend it can end up putting multiple different clients' requests on the same connection. The moment there are two ways to refer to a service (e.g., global vs regional name) you see this change per-request.

Okay, I think those are both good reasons for this approach.  Sounds good to me.